posts/2022/10-01-api_queries_with_finch.md

```diff
@@ -0,0 +1,173 @@
+
+2. This one is for people who like copying code from other projects or the docs (like me): make sure you pass the correct name into the `request/3` function. That means the name under which you added Finch to your supervision tree.
+
+
+
+# 3. Adding a search param to the query
+
+
+Now that we know how to run a basic query let's look at how we can pass query params, such as search terms, API keys etc., into our API request.
+
+This is the step that tripped me up at first. Coming from HTTPoison, I thought I would have to pass in any query params as a keyword list somewhere in the `build/5` function. Spoiler! This is not the case.
+Instead, we need to append any query params to our base URL. 
+
+For our Chuck Norris API, a valid search URL would, for example, look like this: "https://api.chucknorris.io/jokes/search?query=cat"
+To get cat-related Chuck Norris jokes, we would have to pass the entire URL as the second argument into our `build/5` function.
+
+If our search term is the only param that we are adding to our request, we can probably get away with writing something like this:
+
+```
+def query_api(thing_we_are_searching_for) do
+  Finch.build(:get, "https://api.chucknorris.io/jokes/search?query=#{thing_we_are_searching_for}") |> Finch.request(NewApp.Finch)
+end 
+```
+
+But what if we want to add more than that?
+
+
+# 4. Adding several params 
+
+
+When we are querying APIs, we will often have to add more than just a search term to our query string. A lot of APIs require an API key. Or we might be able to set a limit for how many results we want or filter the results based on a timestamp. 
+Of course, we could just extend the `query_api/1` function above and pass in all the other params. 
+But that would make for some really ugly code (at least in my opinion). So let's stay away from that.
+
+Let's take a more dynamic approach instead!
+
+I'm currently big into writing helper modules, so let's put our query builder functions into a helper module.
+
+```
+defmodule NewApp.FinchHelpers do
+  @moduledoc """
+  Functions for building an API call through Finch
+  """
+  @type params :: %{api_key: String.t() | nil, q: String.t() | nil, limit: pos_integer} | []
+  
+
+  @spec build_query(params(), String.t()) :: Finch.Request.t()
+  def build_query(params, url) do
+    url
+    |> append_params(params)
+    |> String.trim_trailing("&")
+    |> then(fn x -> Finch.build(:get, x) end)
+  end
+
+  def append_params(query, []) do
+    query
+  end
+
+  def append_params(query, params) do
+    Enum.reduce(params, "#{query}?", fn {k, v}, acc -> acc <> "#{k}=#{v}&" end)
+  end
+```
+
+In this example, we are adding an API key, a query term (q) and a limit to our URL. All the params are passed into our main `build_query/2` function in a map so that both keys and values can be used for building the query.
```

Comment:
```suggestion
In this example, we are adding an API key, the query variable `q` and a limit to our URL. All the parameters are passed into our main `build_query/2` function in a map so that both keys and values can be used for building the query.
```